pkg/kube/cluster-init.sh

```diff
@@ -498,21 +508,41 @@ if [ ! -f /var/lib/all_components_initialized ]; then
         fi
 
         if [ ! -f /var/lib/longhorn_initialized ]; then
-                wait_for_item "longhorn"
-                logmsg "Installing longhorn version ${LONGHORN_VERSION}"
-                apply_longhorn_disk_config "$HOSTNAME"
-                lhCfgPath=/var/lib/lh-cfg-${LONGHORN_VERSION}.yaml
-                if [ ! -e $lhCfgPath ]; then
-                        curl -k https://raw.githubusercontent.com/longhorn/longhorn/${LONGHORN_VERSION}/deploy/longhorn.yaml > "$lhCfgPath"
+                if [ ! -f /var/lib/longhorn_installing ]; then
+                        wait_for_item "longhorn"
+                        logmsg "Installing longhorn version ${LONGHORN_VERSION}"
+                        apply_longhorn_disk_config "$HOSTNAME"
+                        lhCfgPath=/var/lib/lh-cfg-${LONGHORN_VERSION}.yaml
+                        if [ ! -e $lhCfgPath ]; then
+                                curl -k https://raw.githubusercontent.com/longhorn/longhorn/${LONGHORN_VERSION}/deploy/longhorn.yaml > "$lhCfgPath"
+                        fi
+                        if ! grep -q 'create-default-disk-labeled-nodes: true' "$lhCfgPath"; then
+                                sed -i '/  default-setting.yaml: |-/a\    create-default-disk-labeled-nodes: true' "$lhCfgPath"
+                        fi
+                        kubectl apply -f "$lhCfgPath"
+                        touch /var/lib/longhorn_installing
                 fi
-                if ! grep -q 'create-default-disk-labeled-nodes: true' "$lhCfgPath"; then
-                        sed -i '/  default-setting.yaml: |-/a\    create-default-disk-labeled-nodes: true' "$lhCfgPath"
+                lhStatus=$(kubectl -n longhorn-system get daemonsets -o json | jq '.items[].status | .numberReady==.desiredNumberScheduled' | tr -d '\n')
+                if [ "$lhStatus" = "truetruetrue" ]; then
+                        logmsg "longhorn ready"
+                        rm /var/lib/longhorn_installing
+                        touch /var/lib/longhorn_initialized
                 fi
-                kubectl apply -f "$lhCfgPath"
-                touch /var/lib/longhorn_initialized
         fi
 
-        if [ -f /var/lib/k3s_initialized ] && [ -f /var/lib/kubevirt_initialized ] && [ -f /var/lib/longhorn_initialized ]; then
+        #
+        # Descheduler
+        #
+        if [ ! -f /var/lib/descheduler_initialized ]; then
+                wait_for_item "descheduler"
+                logmsg "Installing Descheduler"
+                DESCHEDULER_VERSION="v0.29.0"
+                kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/descheduler/${DESCHEDULER_VERSION}/kubernetes/base/rbac.yaml
+                kubectl apply -f /etc/descheduler-policy-configmap.yaml
+                touch /var/lib/descheduler_initialized
```

Comment:
I think this works, but would it be better to use the actual state, rather than some artificial external flag?

Actually, why bother at all? You can apply the same manifest over and over, and if nothing has changed, it will be a noop. Why not just download the rbac manifest locally, and just apply it no matter what?

Reply:
It's a good point, somehow we ended up down the path of setting these flags but I don't remember why.  I guess the flags reduce some kube api and etcd load and replace it with a tiny local filesystem read but they do duplicate state which is already in the db.  

I guess we could switch over to re-applying the manifest but I might suggest we move that to a separate slower loop at once every 5min(or longer) instead of every 15 seconds which the main loop runs at.

Reply:
Your call. I don't think even every 15 seconds would be very expensive, but that all is relative. Just be sure you download and save it, rather than applying it off of the URL.

You could do a `kubectl get` to check if it is there, but that is what apply will do under the covers anyways, so why bother?

This retry every 15 seconds is until the whole init is complete, right? After that it stops entirely? So even reapplying every 15 seconds is not bad for a few retries.

IIRC, most cluster initializers (like kubeadm and similar) do a similar thing, just apply manifests over and over until things work. Much easier than trying to figure out what was applied and what wasn't.

Reply:
Maybe I refactored a bit too far, but I rewrote this path to move the install blocks for longhorn and descheduler into longhorn-utils.sh and descheduler-utils.sh (and source each in near the top of cluster-init.sh).  If `kubectl apply` returns an error for either then continue the main loop and retry next time.  This also stops using a flag file for either of them, the remaining flag files for multus and kubevirt can be refactored in a different PR (since this is already growing a bit).

Reply:
@deitch I'll mark this resolved, please re-open if that's not the case.